File: programmers/코딩테스트 고득점 Kit/이분탐색/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
typedef long long ll;
bool Possible(ll val, int n, vector<int> times) {
	int index = 0;
	while (index < times.size()) {
		n -= (val / times[index]);
		index++;
		if (n <= 0) return true;
	}
	return false;
}
ll BinarySearch(int n, vector<int> times) {
	ll left = 0, right = (ll)n * (ll)times[0], ans;
	while (left <= right) {
		ll mid = (left + right) / 2;
		if (Possible(mid, n, times)) ans = mid, right = mid - 1;
		else left = mid + 1;
	}
	return ans;
}
long long solution(int n, vector<int> times) {
	sort(times.begin(), times.end());
	long long answer = BinarySearch(n, times);
	return answer;
}

File: programmers/코딩테스트 고득점 Kit/이분탐색/입국심사.java
class Solution {
    public static final long MAX_TOTAL_TIME = 1000000000000000000L;
    public boolean canExamineTime(long examineTime, int[] times, int n) {
        long count = 0;
        for(int time : times) {
            count += examineTime / time;
        }
        if(count >= n) return true;
        else return false;
    }
    
    public long solution(int n, int[] times) {
        long answer = MAX_TOTAL_TIME;
        long left = 0, right = MAX_TOTAL_TIME;
        while(left <= right) {
            long mid = (left + right) / 2;
            if(canExamineTime(mid, times, n)) {
                right = mid - 1;
                answer = mid;
            }
            else left = mid + 1;
        }
        return answer;
    }
}